src/js/components/FrameworkConfiguration.js

```diff
@@ -0,0 +1,387 @@
+        .split("_")
+        .map(word => word[0].toUpperCase() + word.slice(1))
+        .join(" ");
+
+      renderKeys[key] = formattedKey;
+      this.getHashMapRenderKeys(formData[key], renderKeys);
+    });
+  }
+
+  handleJSONToggle() {
+    this.setState({ jsonEditorActive: !this.state.jsonEditorActive });
+  }
+
+  handleCloseConfirmModal() {
+    this.setState({ isConfirmOpen: false });
+  }
+
+  handleGoBack() {
+    const { reviewActive, hasChangesApplied } = this.state;
+
+    if (reviewActive && hasChangesApplied) {
+      this.setState({ reviewActive: false });
+    } else if (hasChangesApplied) {
+      this.setState({ isConfirmOpen: true });
+    } else {
+      this.handleConfirmGoBack();
+    }
+  }
+
+  handleServiceReview() {
+    const { handleRun } = this.props;
+    const { reviewActive } = this.state;
+
+    if (reviewActive) {
+      handleRun();
+    } else {
+      this.setState({ reviewActive: true });
+    }
+  }
+
+  handleConfirmGoBack() {
+    const { handleGoBack } = this.props;
+
+    this.setState({ isConfirmOpen: false, isOpen: false }, () => {
+      // Once state is set, start a timer for the length of the animation and
+      // navigate away once the animation is over.
+      setTimeout(handleGoBack, 300);
+    });
+  }
+
+  getSecondaryActions() {
+    const { reviewActive, hasChangesApplied } = this.state;
+
+    return [
+      {
+        className: "button-stroke",
+        clickHandler: this.handleGoBack.bind(this),
+        label: reviewActive && hasChangesApplied ? " Back" : "Cancel"
+      }
+    ];
+  }
+
+  getPrimaryActions() {
+    const { jsonEditorActive, reviewActive, hasChangesApplied } = this.state;
+    const { deployErrors, formErrors } = this.props;
+
+    let label = "Review & Run";
+    if (reviewActive) {
+      label = "Run Service";
+    }
+
+    const hasFormErrors = Object.keys(formErrors).reduce(
+      (sum, tab) => sum + formErrors[tab],
+      0
+    );
+
+    const actions = [
+      {
+        className: "button-primary flush-vertical",
+        clickHandler: this.handleServiceReview.bind(this),
+        label,
+        disabled: hasFormErrors ||
+          (reviewActive && deployErrors && !hasChangesApplied)
+      }
+    ];
+
+    if (!reviewActive) {
+      actions.unshift({
+        node: (
+          <ToggleButton
+            className="flush"
+            checkboxClassName="toggle-button toggle-button-align-left"
+            checked={jsonEditorActive}
+            onChange={this.handleJSONToggle.bind(this)}
+            key="json-editor"
+          >
+            JSON Editor
+          </ToggleButton>
+        )
+      });
+    }
+
+    return actions;
+  }
+
+  getPageContents() {
+    const {
+      reviewActive,
+      jsonEditorActive,
+      focusField,
+      activeTab
+    } = this.state;
+    const {
+      packageDetails,
+      deployErrors,
+      formErrors,
+      isInitialDeploy,
+      formData
+    } = this.props;
+
+    if (!reviewActive) {
```

Comment:
Could we negate this condition and put the code for the review screen into a helper function like `getReviewScreen`. This would increase the readability of this function.

Reply:
Yes, this will be much nicer. Thanks, will do this.